### PR TITLE
Ensure validation occurs on clone addresses too

### LINKED ISSUE #14984: Mirror update does not respect restrictions
- Gitea version (or commit ref): current master

## Description

If repository migration is restricted with `setting.Migrations.AllowedDomains`, `setting.ImportLocalPaths` or `setting.Migrations.AllowLocalNetworks` these are [only enforced at the initial migration](https://github.com/go-gitea/gitea/blob/cc6470425c720a05abc55b21f65e3d8179618900/modules/migrations/migrate.go#L37). After the first migration you can change the mirror url to any disallowed domain. The settings page also doesn't work with local paths because `The provided url is invalid. Only http(s):// or git:// locations can be mirrored from.` which is not true. Both modules should share the same logic.

An other question: Should a mirror sync be cancled if the original clone url gets disallowed afterwards?

## Screenshots

grafik

### PR DESCRIPTION
**Pull request by zeripath**
Fix #14984

Signed-off-by: Andrew Thornton <art27@cantab.net>

### CODE REVIEW

**Review by lafriks — APPROVED** [review]

File: modules/migrations/migrate.go [hunk 1]
```diff
@@ -34,39 +36,60 @@ func RegisterDownloaderFactory(factory base.DownloaderFactory) {
...
+			return err
+		}
+		if !isDir {
+			return &models.ErrInvalidCloneAddr{Host: "<LOCAL_FILESYSTEM>", IsInvalidPath: true}
 		}
+
+		return nil
+	}
+
+	if u.Scheme == "git" && u.Port() != "" && (strings.Contains(remoteURL, "%0d") || strings.Contains(remoteURL, "%0a")) {
```
Comment (noerw) [inline modules/migrations/migrate.go:67]: This seems subtle, might make sense to separate this condition into its own func so it's less likely to occur again that checks for the same thing are implemented differently

File: modules/migrations/migrate_test.go [hunk 2]
```diff
@@ -16,30 +16,30 @@ func TestMigrateWhiteBlocklist(t *testing.T) {
 	setting.Migrations.AllowedDomains = []string{"github.com"}
 	assert.NoError(t, Init())
 
-	err := isMigrateURLAllowed("https://gitlab.com/gitlab/gitlab.git")
+	err := IsMigrateURLAllowed("https://gitlab.com/gitlab/gitlab.git", nil)
```
Comment (noerw) [inline modules/migrations/migrate_test.go]: Maybe also add tests for behaviour when doer is set?

**Review by noerw — APPROVED** [review]
2 nits, but otherwise this looks good.
(re your comment; I couldn't test on windows)
also, test fail is related

**Comment by zeripath** [comment]
I'm not sure what's happening on these tests because they work locally...

**Comment by noerw** [comment]
for me `make test` fails locally

File: modules/migrations/migrate.go (hunk 1 above)
Comment (zeripath) [reply modules/migrations/migrate.go:67]: This was in its own function that wasn't being used! The point is to make sure that there really is only place where this is done.

**Comment by techknowlogick** [comment]
🚀

**Review by 6543 — COMMENTED** [review]
thanks for the fix & cleanup!

**Pull request merged**